Add Docker Hub login to Drone's Kubernetes pipelines

wadells opened the pull request:
## Summary

After moving Drone to AWS, we're seeing image pulls get rate limited because they're all coming from the same IP (an AWS NAT gateway). To avoid the rate limiting on AWS, we refactor pipelines to cache/reuse images where possible, as well as add authentication to Docker Hub pulls.

## Related Issues & PRs

Supersedes https://github.com/gravitational/teleport/pull/23831

Contributes to https://github.com/gravitational/SecOps/issues/285

These credentials were added to Drone in:
* GCP https://github.com/gravitational/ops/pull/485
* AWS https://github.com/gravitational/cloud-terraform/pull/1848

Backports:
* https://github.com/gravitational/teleport/pull/23958
* https://github.com/gravitational/teleport/pull/23959
* https://github.com/gravitational/teleport/pull/23972
* https://github.com/gravitational/teleport/pull/23973

## Notes
Most credit goes to @tcsc for this patch.  This is his work from #23831, with the testing changes rebased out and a race condition fixed (see the later commits).

GCP did not have this issue for a variety of reasons:
1) Each node had their own IP -- they weren't on private subnets.  These IPs were rotated regularly as the pool scaled up and down, getting arbitrary IPs from GCPs pools.
2) GKE has caching for frequently used images that may have eliminated some traffic to Docker Hub: https://cloud.google.com/container-registry/docs/pulling-cached-images

## Testing
This is undergoing final testing at:

* build: https://drone.platform.teleport.sh/gravitational/teleport-private/515
* promote: https://drone.platform.teleport.sh/gravitational/teleport-private/534

There are some failures latent in master (related to tenant-operator + arm64 + sqlite) that I expect the above to hit.  As long as we're not hitting Docker Hub rate limits, we're good to merge.

tcsc committed:
Add Docker Hub login to kubernetes pipelines

After moving Drone to AWS, we're seeing image pulls get rate limited
because they're all coming from the same IP (an AWS NAT gateway).

To avoid this, we refactor pipelines to cache/reuse images where
possible, as well as add authentication to dockerhub pulls.

reedloden commented on dronegen/common.go:
```diff
@@ -86,6 +86,25 @@ var (
+	// volumeDockerConfig is a temporary volume for storing docker
+	// credentials for use with the Docker-in-Docker service we use
+	// to isolate the host machines docker daemon from the one used
+	// during the build. Mount this any tome you use `volumeDocker`
```
```suggestion
	// during the build. Mount this any time you use `volumeDocker`
```

wadells commented on dronegen/main.go:
```diff
@@ -39,6 +39,22 @@ func main() {
+	// the named secret as its credentials when pulling images from
+	// dockerhub.
+	//
+	// Exec pipelines to not have the `image_pull_secrets` option, as
```
```suggestion
	// Exec pipelines do not have the `image_pull_secrets` option, as
```

wadells committed:
Drop dockerVolumes and dockerVolumeRefs

We don't actually consistently want these in all places.  E.g. parallel
pipelines cannot share a volumeRefDockerConfig, as they'll stop on each
others login information.

wadells committed:
Remove shared docker config from parallel pipelines

A shared volume results in the different steps racing against each
other.

wadells committed:
Remove docker config from relcli steps

We don't actually pull from dockerhub in these steps.

tcsc committed:
Fix typos

Co-authored-by: Reed Loden <reed@goteleport.com>
Co-authored-by: Walt <walt@goteleport.com>

wadells commented:
The test builds aren't green, but that is because of issues latent in head

* apple build failures 
* a tenant-operator build failure that @hugoShaka is chasing down

The exact same fixes were green in the v12 port, so I'm going to call that good enough.
